modules/gapi/include/opencv2/gapi/s11n.hpp

```diff
@@ -51,8 +51,60 @@ cv::GRunArgs deserialize(const std::vector<char> &p) {
     return detail::getRunArgs(p);
 }
 
-
 } // namespace gapi
 } // namespace cv
 
+namespace cv {
+namespace gimpl {
+namespace s11n {
+namespace I {
+    struct GAPI_EXPORTS OStream {
+        virtual ~OStream() = default;
+
+        // Define the native support for basic C++ types at the API level:
+        virtual OStream& operator<< (bool) = 0;
+        virtual OStream& operator<< (char) = 0;
+        virtual OStream& operator<< (unsigned char) = 0;
+        virtual OStream& operator<< (short) = 0;
+        virtual OStream& operator<< (unsigned short) = 0;
+        virtual OStream& operator<< (int) = 0;
+        virtual OStream& operator<< (uint32_t) = 0;
+        virtual OStream& operator<< (uint64_t) = 0;
+        virtual OStream& operator<< (float) = 0;
+        virtual OStream& operator<< (double) = 0;
+        virtual OStream& operator<< (const std::string&) = 0;
+    };
+
+    struct GAPI_EXPORTS IStream {
+        virtual ~IStream() = default;
+
+        virtual IStream& operator>> (bool &) = 0;
+        virtual IStream& operator>> (std::vector<bool>::reference) = 0;
+        virtual IStream& operator>> (char &) = 0;
+        virtual IStream& operator>> (unsigned char &) = 0;
+        virtual IStream& operator>> (short &) = 0;
+        virtual IStream& operator>> (unsigned short &) = 0;
+        virtual IStream& operator>> (int &) = 0;
+        virtual IStream& operator>> (float &) = 0;
+        virtual IStream& operator>> (double &) = 0;
+        virtual IStream& operator >> (uint32_t &) = 0;
+        virtual IStream& operator >> (uint64_t &) = 0;
+        virtual IStream& operator>> (std::string &) = 0;
+    };
+} // namespace I
+
```

Comment:
I'd also add here our generic (template) s11n overloads for vectors, maps, etc. There's at least one user for `map<>` :)

Reply:
It won't be pure interface in that case since templates are not allowed with `virtual`. Added some types though

Reply:
It will be pure interface. :)